### DISCUSSION TITLE
runtime reflection of table/schema name on event DDL execution

### DISCUSSION THREAD

**Discussion by mattp-** [sqlalchemy/sqlalchemy#11846] [Usage Questions]
Hi,
This may be possible through some wizardry/confluence of features, but i'm struggling to figure it out. I'm generating partitioning on my model doing something like [zzzeek](https://github.com/zzzeek) shared in the recipe here:

https://github.com/sqlalchemy/sqlalchemy/issues/5313#issuecomment-931332837

However, I have an additional wrinkle in that I am trying to drive database parameters, including the schema, as config, making it a runtime set parameter / not defined in the model code. I am currently using declarative base / models. effectively I want/need the schema that is set as a table_arg or with MetaData at runtime, after the models are loaded. The problem is the DDL as written in the shared recipe is hardcoded. How can I dynamically render the table & schema if set?

I see https://docs.sqlalchemy.org/en/20/core/ddl.html#sqlalchemy.schema.DDL defines context vars for table/schema/fullname if acting on a table name, but oddly when https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/sql/compiler.py#L6576 is reached for this code, ddl.target is None.

 I am using schema_translate_map to achieve runtime mapping of schemas - this works fine for the actual table creation DDL, but it's not clear to me how to refer to the table objects in the DDL that make it pass through schema_translate_map. If there is a smarter way to do this I'd be very open to it.

Thank you!

**Comment by CaselIT** [comment]
Hi,

Can you provide an example of what does not work? With schema_translate_map I think you would use the schema token there, not the table one

**Reply by mattp-** [reply]
basically the example as provided, but with a reference to schema. what im trying to do is make an invocation to pgpartman.create_parent, which is a function that requires the explicit named qualified table name with schema:

```
    @classmethod
    def __declare_last__(cls):
        ddl = DDL("SELECT pgpartman.create_parent( p_parent_table := '%(schema)s.salt_events', p_control := 'alter_time', p_interval := 'weekly', p_type := 'native', p_constraint_cols := '{{jid}}', p_jobmon := false)")
        event.listen(cls.__table__, "after_create", ddl)
```

**Reply by CaselIT** [reply]
I still don't understand what your issue is. That's why we ask for executable examples.
Take this example that runs correctly and try to reproduce your issue. Once you post the modified script we can check if it's an issue or not.
```py
import sqlalchemy as sa
from sqlalchemy.event import listen
from sqlalchemy.orm import (
    DeclarativeBase,
    Mapped,
    mapped_column,
    configure_mappers,
)

m = sa.MetaData(schema="X")
t = sa.Table("t", m, sa.Column("data", sa.Text))

d = sa.DDL("insert into %(schema)s.%(table)s values ('a'), ('b')")
listen(t, "after_create", d)


options = {"schema_translate_map": {"X": "test_schema"}}
engine = sa.create_engine(
    "postgresql://scott:tiger@localhost:5432/test", echo=True
)
with engine.execution_options(**options).connect() as c:
    m.create_all(c)


class Base(DeclarativeBase):
    metadata = sa.MetaData(schema="X")

class Tbl(Base):
    __tablename__ = "foo"
    id: Mapped[int] = mapped_column(primary_key=True)
    data: Mapped[str]

    @classmethod
    def __declare_last__(cls):
        print("<here>")
        d = sa.DDL(
            "insert into %(schema)s.%(table)s(data) values ('%(schema)s.%(table)s'), ('b')"
        )
        listen(cls.__table__, "after_create", d)

configure_mappers()

with engine.execution_options(**options).connect() as c:
    Base.metadata.create_all(c)
with engine.connect() as c:
    Base.metadata.create_all(c.execution_options(**options))
```

**Comment by mattp-** [comment]
im unable to reproduce this now that i've tried working on creating an isolated example; I'm going to close this for now. thank you!

**Reply by CaselIT** [reply]
feel free to reopen if you find a way to reproduce it!
